x-pack/plugins/maps/public/shared/layers/sources/client_file_source/geojson_file_source.js

```diff
@@ -0,0 +1,120 @@
+import { GEOJSON_FILE } from '../../../../../common/constants';
+import { ClientFileCreateSourceEditor } from './create_client_file_source_editor';
+import { ESSearchSource } from '../es_search_source';
+import uuid from 'uuid/v4';
+import _ from 'lodash';
+
+export class GeojsonFileSource extends AbstractVectorSource {
+
+  static type = GEOJSON_FILE;
+  static title = 'Import GeoJSON Vector File';
+  static description = 'Upload a GeoJSON file & Index in Elasticsearch';
+  static icon = 'emsApp';
+  static indexReadyFile = true;
+
+  static createDescriptor(featureCollection, name) {
+    return {
+      type: GeojsonFileSource.type,
+      featureCollection,
+      name
+    };
+  }
+
+  static viewIndexedData = (addAndViewSource, inspectorAdapters, onIndexSuccess) => {
+    return ({ fields, id, success }) => {
+      if (!success) {
+        console.error('Unable to view indexed data');
+        return;
+      }
+      const geoFieldArr = fields.filter(
+        field => ['geo_point', 'geo_shape'].includes(field.type)
+      );
+      const geoField = _.get(geoFieldArr, '[0].name');
+      const indexPatternId = id;
+      if (!indexPatternId || !geoField) {
+        addAndViewSource(null);
+      } else {
+        const source = new ESSearchSource({
+          id: uuid(),
+          indexPatternId,
+          geoField,
+        }, inspectorAdapters);
+        addAndViewSource(source);
+        onIndexSuccess();
+      }
+    };
+  };
+
+  static previewGeojsonFile = (onPreviewSource, inspectorAdapters) => {
+    return (geojsonFile, name) => {
+      if (!geojsonFile) {
+        onPreviewSource(null);
+        return;
+      }
+      const sourceDescriptor = GeojsonFileSource.createDescriptor(geojsonFile, name);
+      const source = new GeojsonFileSource(sourceDescriptor, inspectorAdapters);
+      onPreviewSource(source);
+    };
+  };
+
+  static renderEditor({ onPreviewSource, inspectorAdapters, addAndViewSource,
+    boolIndexData, onRemove, onIndexReadyStatusChange, setFailures, onIndexSuccess }) {
+    return (
+      <ClientFileCreateSourceEditor
+        previewGeojsonFile={
+          GeojsonFileSource.previewGeojsonFile(
+            onPreviewSource,
+            inspectorAdapters
+          )
+        }
+        viewIndexedData={
+          GeojsonFileSource.viewIndexedData(
+            addAndViewSource,
+            inspectorAdapters,
+            onIndexSuccess
+          )
+        }
+        boolIndexData={boolIndexData}
+        onRemove={onRemove}
+        onIndexReadyStatusChange={onIndexReadyStatusChange}
+        onIndexAddSuccess={indexAddResp => {
+          if (indexAddResp.failures && indexAddResp.failures.length) {
+            setFailures && setFailures(indexAddResp.failures);
+          }
+        }}
+      />
+    );
+  }
+
+  async getGeoJsonWithMeta() {
+    return {
+      data: this._descriptor.featureCollection,
+      meta: {}
+    };
+  }
+
+  async getDisplayName() {
+    try {
```

Comment:
this try-catch seems dysfunctional. each source should have a `name`, because the `createDescriptor` enforces it. and if it doesn't, this won't throw because each and every object under `shared/` must be instantiated with a descriptor